lib/network/rest/client.js

```diff
@@ -0,0 +1,267 @@
+const Logger = require('werelogs').Logger;
+const constants = require('../../constants');
+const utils = require('./utils');
+
+function setRequestUids(reqHeaders, reqUids) {
+    // inhibit 'assignment to property of function parameter' -
+    // this is what we want
+    // eslint-disable-next-line
+    reqHeaders['X-Scal-Request-Uids'] = reqUids;
+}
+
+function setRange(reqHeaders, range) {
+    const rangeStart = range[0] !== undefined ? range[0].toString() : '';
+    const rangeEnd = range[1] !== undefined ? range[1].toString() : '';
+    // inhibit 'assignment to property of function parameter' -
+    // this is what we want
+    // eslint-disable-next-line
+    reqHeaders['Range'] = `bytes=${rangeStart}-${rangeEnd}`;
+}
+
+function setContentType(reqHeaders, contentType) {
+    // inhibit 'assignment to property of function parameter' -
+    // this is what we want
+    // eslint-disable-next-line
+    reqHeaders['Content-Type'] = contentType;
+}
+
+function setContentLength(reqHeaders, size) {
+    // inhibit 'assignment to property of function parameter' -
+    // this is what we want
+    // eslint-disable-next-line
+    reqHeaders['Content-Length'] = size.toString();
+}
+
+function makeErrorFromHTTPResponse(response) {
+    const rawBody = response.read();
+    const body = (rawBody !== null ? rawBody.toString() : '');
+    let error;
+    try {
+        const fields = JSON.parse(body);
+        error = new Error(fields.message);
+        Object.assign(error, fields);
+    } catch (err) {
+        error = new Error(body);
+    }
+    error.code = response.statusCode;
+    error.remote = true;
+    return error;
+}
+
+
+/**
+ * @class
+ * @classdesc REST Client interface
+ *
+ * The API is usable when the object is constructed.
+ */
+class RESTClient {
+    /**
+     * Interface to the data file server
+     * @constructor
+     * @param {Object} params - Contains the basic configuration.
+     * @param {String} params.host - hostname or ip address of the
+     *   RESTServer instance
+     * @param {Number} params.port - port number that the RESTServer
+     *   instance listens to
+     * @param {Logger} [params.log] - logging configuration
+     */
+    constructor(params) {
+        assert(params.host);
+        assert(params.port);
+
+        this.host = params.host;
+        this.port = params.port;
+        this.setupLogging(params.log);
+        this.httpAgent = new http.Agent({ keepAlive: true });
+    }
+
+    setupLogging(config) {
+        let options = undefined;
+        if (config !== undefined) {
+            options = {
+                level: config.logLevel,
+                dump: config.dumpLevel,
+            };
+        }
+        this.logging = new Logger('RESTClient', options);
+    }
+
+    createLogger(reqUids) {
+        return reqUids ?
+            this.logging.newRequestLoggerFromSerializedUids(reqUids) :
+            this.logging.newRequestLogger();
+    }
+
+    createRequest(method, headers, key = '') {
```

Comment:
Well, not exactly, actually it creates the fields necessary to issue the request head. So maybe createRequestHeadFields can be a better name.